src/transformers/modeling_tf_pytorch_utils.py

```diff
@@ -285,11 +289,164 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
+    if np.isscalar(array):
+        array = np.array(array)
+    check_valid(array, pt_weight)
+    return array
+
+
+def apply_transpose_reverse(array, transpose, symbolic_weight):
+    """
+    Given a numpy array, a transpose enum, and the target tensor shape Attempt to modify the array into something
+    fitting the target tensor The direction here is PT -> TF
+    """
+    if transpose is TransposeType.CONV2D:
+        # Conv2D weight:
+        #    PT: (num_out_channel, num_in_channel, kernel[0], kernel[1])
+        # -> TF: (kernel[0], kernel[1], num_in_channel, num_out_channel)
+        array = np.transpose(array, axes=(2, 3, 1, 0))
+    elif transpose is TransposeType.CONV1D:
+        # Conv1D weight:
+        #    PT: (num_out_channel, num_in_channel, kernel)
+        # -> TF: (kernel, num_in_channel, num_out_channel)
+        array = np.transpose(array, axes=(2, 1, 0))
+    elif transpose is TransposeType.SIMPLE:
+        array = np.transpose(array)
+
+    if len(symbolic_weight.shape) < len(array.shape):
+        array = np.squeeze(array)
+    elif len(symbolic_weight.shape) > len(array.shape):
+        array = np.expand_dims(array, axis=0)
+
+    if list(symbolic_weight.shape) != list(array.shape):
+        squeezed_tensor_shape = [dim for dim in symbolic_weight.shape if dim != 1]
+        squeezed_array_shape = [dim for dim in array.shape if dim != 1]
+        if squeezed_array_shape == squeezed_tensor_shape:
+            array = np.reshape(array, symbolic_weight.shape)
+
+    check_valid(array, symbolic_weight)
+
+    return array
+
+
+def check_valid(array, tensor):
+    """Make sure the modified tensor and the target tensor match, raises ValueError if they don't"""
+    if tensor.shape != array.shape:
+        raise ValueError(f"We could not match the dimensions of tensors {tensor.shape} vs {array.shape}")
+
+
 def load_tf2_checkpoint_in_pytorch_model(pt_model, tf_checkpoint_path, tf_inputs=None, allow_missing_keys=False):
     """
     Load TF 2.0 HDF5 checkpoint in a PyTorch model We use HDF5 to easily do transfer learning (see
     https://github.com/tensorflow/tensorflow/blob/ee16fcac960ae660e0e4496658a366e2f745e1f0/tensorflow/python/keras/engine/network.py#L1352-L1357).
     """
+    try:
+        return _load_tf2_checkpoint_in_pytorch_model(
+            pt_model, tf_checkpoint_path, tf_inputs=tf_inputs, allow_missing_keys=allow_missing_keys
+        )
+    except Exception as e:
+        logger.warning(
+            "We were not successful in loading the model from TF without loading the full model. This is NOT NORMAL, please warn the devs about this https://github.com/huggingface/transformers/issues/new?assignees=&labels=&template=bug-report.md&title= . Defaulting to legacy code"
+        )
+        logger.error(f"Error was {e}")
+        return legacy_load_tf2_checkpoint_in_pytorch_model(
+            pt_model, tf_checkpoint_path, tf_inputs=tf_inputs, allow_missing_keys=allow_missing_keys
+        )
+
```

Comment:
This is the new code with `try..except` that should prevent catastophic failures.

@sgugger @LysandreJik Are you ok with the wording and how it works?

This PR is now adding code only, so you can just check the new code (and verifying `legacy_load...` is the same as the old `load_tf2_...`